deepreg/model/layer.py

```diff
@@ -665,3 +666,123 @@ def call(self, inputs, training=None, **kwargs):
         r2 = self._conv3d_block(inputs=h0, training=training)
         h1 = self._residual_block(inputs=[r2, r1], training=training)
         return h1
+
+
+class BSplines3DTransform(tf.keras.layers.Layer):
+    """
+     Layer for BSplines interpolation with precomputed cubic spline filters.
+     It assumes a full sized image from which: (1) it compute the contol points values by downsampling the initial
+     image (2) performs the interpolation and (3) crops the image around the valid values.
+
+    :param cp_spacing: _int_ or tuple of three _ints_ specifying the spacing (in pixels) in each dimension.
+                      When a single _int_ is used, the same spacing to all dimensions is used
+    :param kwargs:
+    """
+
+    def __init__(self, cp_spacing: (int, tuple), **kwargs):
+
+        super().__init__(**kwargs)
+
+        if isinstance(cp_spacing, int):
+            self.cp_spacing = (cp_spacing, cp_spacing, cp_spacing)
+        else:
+            self.cp_spacing = cp_spacing
+        self.filters = []
+
+    def build(self, input_shape):
+        """
+        :param input_shape: tuple with the input shape
+        :return:
+        """
+
+        super().build(input_shape)
+
+        b = {
+            0: lambda u: np.float32((1 - u) ** 3 / 6),
+            1: lambda u: np.float32((3 * (u ** 3) - 6 * (u ** 2) + 4) / 6),
+            2: lambda u: np.float32((-3 * (u ** 3) + 3 * (u ** 2) + 3 * u + 1) / 6),
+            3: lambda u: np.float32(u ** 3 / 6),
+        }
+
+        filters = np.zeros(
+            (
+                4 * self.cp_spacing[0],
+                4 * self.cp_spacing[1],
+                4 * self.cp_spacing[2],
+                3,
+                3,
+            ),
+            dtype=np.float32,
+        )
+
+        for u in range(self.cp_spacing[0]):

```

Comment:
these for loops scared me a bit... do we have a better way to do so?

I think this example might inspire you a bit?

https://stackoverflow.com/questions/798854/all-combinations-of-a-list-of-lists

Reply:
I simplified the loops, take a look at the updated file

Reply:
i don't think we can eyeball this to say if it is correctly implemented. Is it possible to use, say sitk.BSplineTransform in the test?